torchtune/training/quantization.py

```diff
@@ -127,6 +152,80 @@ def quantize(self, model):
 ] = enable_8da4w_fake_quant_module_swap
 
 
+class Int8MixedPrecisionTrainingQuantizer:
+    """Apply INT8 mixed-precision training. This only affects weights of ``nn.Linear``
+    modules. During training, weights and activations are dynamically quantized to INT8
+    to utilize fast matrix multiplication with INT8 tensor cores. This is also done in
+    the backward pass.
+
+    The expected end2end speedup is 40% on a single A100 and 70% on a single 4090, with
+    minimal accuracy loss. If convergence is an issue, please refer to torchao
+    documentation below.
+
+    For more details, as well as details about arguments of this quantizer, please refer to
+    https://github.com/pytorch/ao/tree/main/torchao/prototype/quantized_training#int8-mixed-precision
+
+    Args:
+        output (bool): whether to apply INT8 mixed-precision for calculating output.
+        grad_input (bool): whether to apply INT8 mixed-precision for calculating grad_input.
+        grad_weight (bool): whether to apply INT8 mixed-precision for calculating grad_weight.
+
+    Raises:
+        RuntimeError: If runtime requirements for INT8 mixed-precision training are not met.
+
+    NOTE: Due to the limitations of the current implementation, the following
+    requirements must be satisfied to enjoy the expected speedup:
+
+    1. Must use ``torch.compile()`` (set ``compile=True``).
+    2. Inputs to the model must not be too dynamic. For example, when input tokens
+    length changes for every batch, you won't see the expected speedup.
+
+    To satisfy (2), you can use :class:`~torchtune.datasets.PackedDataset` (set
+    ``dataset.packed=True`` and ``tokenizer.max_seq_len`` to a desired value.), which
+    ensures input tokens always have fixed length.
+    """
+
+    def __init__(
+        self,
+        output: bool = True,
+        grad_input: bool = True,
+        grad_weight: bool = True,
```

Comment:
I dont think that there is a way around us always having our own wrapper, since we need it to deal with LoRA/other details. Therefore, I think that we should add "enabled" here, and have config like this:

```
mixed_precision:
	_component_: torchtune.training.quantization.Int8MixedPrecisionTraining
	arg1: True
	arg2: True
	enabled: False
```

That way the configs can advertise this functionality.

I also think that we should rename quantizer to "mixed_precision" or "mixed_precision_training", so its self-explanatory.

Let me know if you need help with making the changes or if you disagree with the ideas above.

thanks again for the PR! @ebsmothers might want to leave a few comments too
For me, after these changes and the test pass, i am good to merge.

Reply:
Yeah personally I like the change to rename `quantizer` -> `mixed_precision` (unless the idea is to later support full low-precision training using the same API, in which case it wouldn't make sense). I am on the fence about the inclusion of the `enabled` flag.. it's nice because it allows us to expose directly in our configs without actually enabling it by default. At the same time, I don't love the pattern of defining dummy objects with corresponding thin wrappers that optionally return them (unless it's absolutely necessary, as is kinda the case with some of our context-manager-based features). Personally I am OK with just adding something like this to all our configs if we wanna advertise this feature:

```
# Set to torchtune.training.Int8MixedPrecisionTrainingQuantizer to use int8 mixed precision
mixed_precision: null
```

Reply:
Yea I agree `quantizer` name right now can sound confusing, since the weight is actually not quantized (well QAT also does not actually quantize the weight :smile:)

Regarding `mixed_precision` name, I'm thinking about what other training recipes out there that would fall under this category (because it would be weird if we have `mixed_precision` but only 1 option). There are FP8 and FP16/BF16 mixed-precision training, but they follow a totally different UX
- **FP8**: the current FP8 training recipes, in torchao or transformer engine, is actually mixed-precision training, since master weights is still in high precision, although people usually don't say it is mixed-precision. Currently FP8 UX is to do module swap (but I think for FP8 all-gather, FP8 also does tensor subclass). I think Vasiliy is having plans to update FP8 UX, not exactly sure. But anyway this is an implementation detail which can be hidden away from torchtune users. (also someone should add FP8 to torchtune :smile:)
- **BF16** (w/ FP32 master weights): for single-GPU, it will be using `torch.autocast(dtype=torch.bfloat16)`. For FSDP2, it would be using `MixedPrecisionPolicy(param_dtype=torch.bfloat16)`. This is very different UX (and mechanism) from INT8 and FP8, and would be hard to unify the torchtune user-facing API (and not sure if torchtune wants to add this option? Btw there can be small gaps between full BF16 training (current torchtune) and BF16 mixed-precision training, especially when LR is small. Would take some efforts for someone to investigate this :sweat_smile:).
- **FP16** (w/ FP32 master weights): probably won't consider this due to complexity with gradient scaling + finetune BF16 models with FP16 is a [bad idea](https://github.com/huggingface/transformers/issues/23918).

There are also MX FP6/FP4 in the new NVIDIA GPUs, but that's too far in the future. Considering that at least we can have INT8 and FP8 mixed-precision training, using the `mixed_precision` name seems reasonable!

Reply:
> also someone should add FP8 to torchtune 

Would it be too different from int8? Or can we leverage the same API and replace the quantization module?
BTW, does FP8 work with FSDP? In this case, is there anything that makes it special, since int8 seems to not be compatible?

Reply:
@felipemello1 Currently torchao API is different https://github.com/pytorch/ao/tree/main/torchao/float8. But it can be hidden away by torchtune (e.g. a separate Quantizer class, or some other name like MixedPrecisionTrainer class)

Both FP8 (by FP8 folks) and INT8 (by me) support FSDP2. The issue is with how torchtune handles distributed weight loading, which is not trivial to handle. FP8 module swap might not have this problem though.